web-ui/src/pages/PermissionsPage.jsx

```diff
@@ -72,21 +73,20 @@ const EditPermissionsPage = () => {
   const [rolePermissions, setRolePermissions] = useState([]);
   const [refresh, setRefresh] = useState(true);
 
-  useEffect(() => {
-    const url = new URL(location.href);
-    const roleName = url.searchParams.get('role');
-    const role = roles.find(r => r.role === roleName);
-    setSelectedRole(role || roles[0]);
-  }, []);
-
-  useEffect(() => {
-    if (!selectedRole) return;
-    const url = new URL(location.href);
-    const params = { role: selectedRole.role };
-    const q = new URLSearchParams(params).toString();
-    const newUrl = url.origin + url.pathname + '?' + q;
-    history.replaceState(params, '', newUrl);
-  }, [selectedRole]);
+  queryParameterSetup([
+    {
+      name: 'role',
+      default: roles[0],
+      value: selectedRole,
+      setter(value) {
+        const role = roles.find(r => r.role === value);
+        setSelectedRole(role);
+      },
+      toQP(selectedRole) {
+        return selectedRole.role;
+      }
```

Comment:
I believe this is the line causing one of the CI tests to fail. Looks like `selectedRole` is sometimes `undefined` when making this call. I see we're grabbing it from state, so I suspect we need to defer QSP set-up until we have the needed state.

Reply:
Great catch! I wonder how I can implement that delay. It's not obvious to me.
I could use `setInterval` and just check repeatedly.
Did you have another approach in mind?